### ISSUE TITLE
Use yield less aggressively in group commit

### ISSUE THREAD

**Issue by agorrod** [wiredtiger/wiredtiger#1591]
Some system schedulers don't de-prioritize yield calls when the number of threads overwhelms the number of CPU cores. The WiredTiger log consolidation code makes extensive use of yield to avoid waiting longer than necessary for a group flush to finish.

It would be better if we backed off using a mechanism other than spinning in a yield loop in ```__wt_log_slot_close```. It has been reported that up to 10x improvement in throughput can be achieved in some MongoDB workloads on some hardware.

**Comment by agorrod** [comment]
A change like:

```
--- a/src/log/log_slot.c
+++ b/src/log/log_slot.c
@@ -213,9 +213,8 @@ retry:
                WT_ASSERT(session, slot->slot_churn >= 0);
        }
 
-       /* Pause to allow other threads a chance to consolidate. */
-       for (yields = slot->slot_churn; yields >= 0; yields--)
-               __wt_yield();
+       /* Sleep in increments of 10th of a millisecond */
+       __wt_sleep(0, slot_churn * 100);
 
        /*
         * Swap out the slot we're going to use and put a free one in the
```

Might be sufficient.

@sueloverso I know we've played with the timing here before, but I don't think I ever tried running with more threads than cores. Maybe we could choose some benchmarks and platforms and try out different combinations?

**Comment by sueloverso** [comment]
I think some of the wtperf configurations use more threads than cores on the AWS boxes.  I think the HDD box only has 8 cores.  In any case, the code you reference is just one of a good number of yield calls in logging.  In particular, I often see a lot of threads in `__wt_log_release` waiting for their turn to call `__wt_write`.  There are also yield calls if the thread needs to wait for the previous log file to close (i.e. if we want to switch from N to N+1, but N-1 still isn't closed).  That is not the typical case though (and we do have a stat for that to know).  If you know of some particular workloads that exhibit problems please let me know.

**Comment by agorrod** [comment]
@RedBeard0531 I opened this issue so that the issue you mentioned in log performance didn't get lost after 3.0

**Comment by RedBeard0531** [comment]
@agorrod FYI, I figured out my computer showed this way more than anyone else's. I was running a ck-patched kernel using the BFS scheduler[1] which is supposed to be tuned for desktop responsiveness. When I rebooted in to a vanilla kernel with the default CFS scheduler I saw significantly less percent of time spent servicing sched_yield. Since most (hopefully all) production servers would be using something other than BFS, I don't think this is as critical of a problem.

We do however need to test on kernels on both sides of 2.6.23 since that is when the default scheduler switched from the O(1) to CFS.

[1] http://en.wikipedia.org/wiki/Brain_Fuck_Scheduler
